### PR TITLE
Add DASC recurrent state sparsity policy

### PR DESCRIPTION
**Pull request by kaix-nv**
### What does this PR do?

Type of change: new feature

Adds an experimental `modelopt.torch.sparsity.state_sparsity` API for the first ModelOpt DASC milestone:

- derives deterministic GDN whole-head decay horizons from `A_log` and `dt_bias` in CPU FP64;
- selects the largest arbitrary positive `Wmax` that passes caller-supplied quality, lifecycle, and physical-storage evidence;
- keeps DASC-NR zero recovery and DASC-WR suffix replay as explicit contracts;
- records model/checkpoint/calibration provenance, masks, gates, and measurements in versioned JSON-safe ModelOpt state;
- validates model structure during restore and rejects stale decay parameters during export/save.

This change is policy-only and does not alter model execution. Runtime ragged checkpoint packing, suffix replay, KDA support, quantization, and GDN/KDA kernels are intentionally out of scope. A serving implementation must preserve convolution state and materialize ordinary dense recurrent state before continuation.

### Usage

```python
import modelopt.torch.sparsity.state_sparsity as mtss

config = {
    "variant": "dasc_wr",
    "wmax_candidates": [32],
    "model_id": "org/model",
    "model_revision": "immutable-revision",
    "model_config_id": "sha256:<config-digest>",
    "calibration_data_id": "sha256:<dataset-and-protocol-digest>",
}

model = mtss.calibrate(model, config, measurements)
deployment_policy = mtss.export_policy(model)
```

`measurements` contains paired dense-versus-DASC quality and physical checkpoint-storage results for every configured candidate. The full schema and example are documented in the sparsity guide.

### Testing

- `PYTHONPATH=$WORKTREE python -m pytest -q tests/unit/torch/sparsity/state_sparsity/test_dasc.py tests/unit/torch/sparsity/weight_sparsity/test_sparsify.py` — 119 passed
- `PYTHONPATH=$WORKTREE python -m pytest -q tests/unit/torch/sparsity/attention_sparsity/test_sparse_attention_mode.py` — 2 passed
- Real-class smoke test with Transformers `Qwen3NextGatedDeltaNet` — horizon analysis, calibration, and policy export passed
- `pre-commit run --files <all changed files>` — all hooks passed
- `git diff --check` — clean

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors).

- Is this change backward compatible?: ✅
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A — no copied code or dependency added
- Did you write any new necessary tests?: ✅
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅
- Did you get Claude approval on this PR?: ❌ — review will be requested after PR creation

### Additional Information

Algorithm reference: [DASC: Decay-Aware State Compression for Hybrid Linear-Attention Serving](https://arxiv.org/abs/2608.30386).

The mode stores calibration evidence rather than treating the weight-derived horizon as sufficient proof. The first runtime milestone should consume this policy in the serving prefix-cache serializer/loader and qualify matched HBM capacity plus prefix-hit latency.

## Summary by CodeRabbit

* **New Features**
  * Added experimental decay-aware recurrent-state sparsity calibration for Gated Delta Net models.
  * Added policy export, restoration metadata, and checkpoint-specific recovery-window selection using quality and storage measurements.
  * Added validation for calibration inputs, model compatibility, decay parameters, policy integrity, and stale metadata.

* **Documentation**
  * Documented configuration, calibration, policy deployment, supported recovery variants, and serving-runtime responsibilities.
  * Clarified that the API exports metadata only and does not modify execution or provide checkpoint-packing and recovery kernels.

### CODE REVIEW

File: modelopt/torch/sparsity/state_sparsity/__init__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Decay-aware sparsity policies for persisted recurrent state."""
+
+from . import mode
+from .api import *
+from .config import *
+from .policy import *
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/__init__.py]: Addressed in signed commit edc91ec0f1 via #2377. The package now uses explicit imports and an explicit `__all__` for the nine public API, schema, and analysis symbols; `mode` remains imported only for registration side effects. A regression test checks that `DASCLayerPolicy` is exported and `mode` is not.

File: modelopt/torch/sparsity/state_sparsity/config.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    "DASCPolicy",
+    "DASCQualityMeasurement",
+]
+
+
+class DASCQualityMeasurement(ModeloptBaseConfig):
+    """Quality and lifecycle measurements for one calibration slice."""
+
+    slice_id: str = Field(min_length=1)
+    perplexity_retention: float = Field(gt=0.0, le=1.0, allow_inf_nan=False)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/config.py]: Addressed in signed commit edc91ec0f1 via #2377. `perplexity_retention` is now any finite positive value. Both the field schema and guide define it as `dense_perplexity / DASC_perplexity = exp(dense_NLL - DASC_NLL)`, so values above one are valid. A test covers 1.0004.

File: modelopt/torch/sparsity/state_sparsity/config.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+from modelopt.torch.opt.config import ModeloptBaseConfig, ModeloptField
+
+__all__ = [
+    "DASCCalibrationMeasurement",
+    "DASCConfig",
+    "DASCPolicy",
+    "DASCQualityMeasurement",
+]
+
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/config.py:33]: Partly addressed in signed commit edc91ec0f1 via #2377: `DASCLayerPolicy` is now exported from both `config.__all__` and the package `__all__`. I am deliberately deferring config and policy deduplication for format v1: the attached and exported policy is a standalone deployment artifact, while the ModelOpt config is mode history, and the current cross-check prevents mixed metadata. Any schema normalization can be done under a future `format_version` without coupling this review fix to a persisted-state migration.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+def _is_gdn_module(module: nn.Module) -> bool:
+    class_name = "".join(
+        character for character in type(module).__name__.lower() if character.isalnum()
+    )
+    return (
+        "gateddeltanet" in class_name
+        and isinstance(getattr(module, "A_log", None), torch.Tensor)
+        and isinstance(getattr(module, "dt_bias", None), torch.Tensor)
+    )
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Addressed in signed commit edc91ec0f1 via #2377. GDN discovery now accepts only the explicit class names `GatedDeltaNet` and `Qwen3NextGatedDeltaNet`, fails closed with a diagnostic listing the supported set, and documents this initial adapter scope. A negative test rejects an otherwise shape-compatible class with a GDN-like name.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        measurements=validated_measurements,
+    )
+
+
+def validate_dasc_model_structure(model: nn.Module, policy: DASCPolicy) -> None:
+    """Reject restoring a policy onto a different GDN module structure."""
+    modules = _get_gdn_modules(model)
+    actual = _canonical_sha256(_model_structure(modules))
+    if actual != policy.model_structure_sha256:
+        raise ApplyModeError("DASC policy does not match the model's GDN module structure")
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Addressed in signed commit edc91ec0f1 via #2377. Structure validation now compares model-derived layer names and head counts against `policy.layers` as well as the structure digest. Deployment validation re-derives every horizon from current decay tensors and policy parameters, then exact-compares the selected retained-head mask. Restore attaches inert metadata before `load_state_dict`, so only structure can be checked at that phase; `export_policy()` performs the complete decay, horizon, and mask check after weights are loaded. Tests cover tampered horizons and a self-consistent but incorrectly recomputed mask.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+
+def update_dasc_metadata(model: nn.Module, config: DASCConfig, metadata: MetadataDict) -> None:
+    """Refresh serialized metadata from the immutable attached DASC policy."""
+    try:
+        policy = DASCPolicy(**getattr(model, _DASC_POLICY_ATTRIBUTE))
+    except (AttributeError, TypeError, ValidationError) as error:
+        raise ApplyModeError("Model has no valid attached DASC policy") from error
+    validate_dasc_model_structure(model, policy)
+    validate_dasc_decay_parameters(model, policy)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py]: Addressed in signed commit edc91ec0f1 via #2377. Save and compose paths now warn and serialize a stale policy rather than raising; `export_policy()` remains fail-closed. Reapplying DASC is allowed so calibration can supersede stale metadata. The digest canonicalizes through BF16 to make routine FP32-to-BF16 casts stable, while export also re-derives horizons and the exact selected mask. Tests cover BF16 casting, non-fatal `modelopt_state()` and `mto.save()`, strict stale export, recalibration, and restore.

File: modelopt/torch/sparsity/state_sparsity/config.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        default=[8, 16, 32, 64, 128, 256],
+        description="Positive candidate windows evaluated during offline calibration.",
+    )
+    min_perplexity_retention: float = ModeloptField(default=0.995)
+    min_top1_agreement: float = ModeloptField(default=0.98)
+    min_checkpoint_savings: float = ModeloptField(default=0.2)
+    model_id: str = ModeloptField(default="", validate_default=True)
+    model_revision: str = ModeloptField(default="", validate_default=True)
+    model_config_id: str = ModeloptField(
+        default="",
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/config.py:137]: Addressed in signed commit edc91ec0f1 via #2377. `DASCConfig` and `DASCPolicy` now use `ConfigDict(extra="forbid", validate_assignment=True, protected_namespaces=())`, matching repository precedent while retaining existing validation behavior.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    metadata.clear()
+    metadata["policy"] = copy.deepcopy(policy.model_dump(mode="json"))
+
+
+def get_attached_dasc_policy(model: nn.Module) -> DASCPolicy:
+    """Return the validated policy attached by conversion or restoration."""
+    try:
+        return DASCPolicy(**getattr(model, _DASC_POLICY_ATTRIBUTE))
+    except (AttributeError, TypeError, ValidationError) as error:
+        raise ApplyModeError("Model has no valid attached DASC policy") from error
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py:168]: Addressed in signed commit edc91ec0f1 via #2377. `get_attached_dasc_policy()` now force-unwraps recognized wrappers, and `update_dasc_metadata()` calls that helper instead of duplicating the attribute and schema logic. A DataParallel export regression test was added.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+def _attach_policy(model: nn.Module, policy: DASCPolicy) -> None:
+    setattr(model, _DASC_POLICY_ATTRIBUTE, policy.model_dump(mode="json"))
+
+
+def convert_dasc_model(
+    model: nn.Module,
+    config: DASCConfig,
+    *,
+    measurements: Iterable[DASCCalibrationMeasurement | dict],
+) -> ConvertReturnType:
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py:52]: Addressed in signed commit edc91ec0f1 via #2377. `measurements` now defaults to `None`, and generic `mto.apply_mode()` raises an actionable `ApplyModeError` directing callers to `state_sparsity.calibrate(...)`. A regression test covers the generic path.
